### ISSUE TITLE
`brew uses mpich2` returns nothing; it should return at least 1 formula

### ISSUE THREAD

**Issue by goxberry** [Homebrew/homebrew#41517]
**tl;dr:** It seems like `brew uses mpich2` should return something, because it's possible to use `mpich2` to build other Homebrew packages that depend on MPI.

**Longer version:** I'm trying to assess what I might need to re-install to migrate from using Open MPI to MPICH, since the two libraries can't coexist easily in Homebrew (see https://github.com/Homebrew/homebrew/issues/22617).

I currently have `open-mpi` installed. To figure out what might need to be reinstalled when migrating to MPICH, I invoked `brew uses open-mpi`, which returns more or less what I'd expect:

```
bash-3.2$ brew uses open-mpi
boost                            homebrew/science/bless           homebrew/science/hdf5            homebrew/science/mumps           homebrew/science/ray             homebrew/science/zoltan
bsponmpi                         homebrew/science/dealii          homebrew/science/hyphy           homebrew/science/neuron          homebrew/science/scalapack       homebrew/versions/boost149
fftw                             homebrew/science/elemental       homebrew/science/hypre           homebrew/science/nonpareil       homebrew/science/scotch          homebrew/versions/boost150
konoha                           homebrew/science/exabayes        homebrew/science/lammps          homebrew/science/p4est           homebrew/science/scotch5         
homebrew/science/abinit          homebrew/science/garli           homebrew/science/lis             homebrew/science/paraview        homebrew/science/slepc           
homebrew/science/abyss           homebrew/science/getdp           homebrew/science/maker           homebrew/science/parmetis        homebrew/science/sundials        
homebrew/science/alpscore        homebrew/science/gmsh            homebrew/science/meme            homebrew/science/pastix          homebrew/science/superlu_dist
homebrew/science/arpack          homebrew/science/gromacs         homebrew/science/mrbayes
```

To figure out what I might be able to install using MPICH, I run `brew uses mpich2`, which returns...nothing. In theory, `homebrew/science/superlu_dist` (https://github.com/Homebrew/homebrew-science/pull/1750), `homebrew/science/hdf5` (https://github.com/Homebrew/homebrew-science/pull/1756), `homebrew/science/sundials` (https://github.com/Homebrew/homebrew-science/pull/1753), `homebrew/science/hypre` (https://github.com/Homebrew/homebrew-science/pull/1752), `homebrew/science/mumps` (https://github.com/Homebrew/homebrew-science/pull/1747), and `boost` (without C++11 support; https://github.com/Homebrew/homebrew/issues/29707) should all come up, at least; this list is not exhaustive.

Should this be happening?

**Comment by mikemcquaid** [comment]
Do you have `homebrew/science` tapped?

**Comment by goxberry** [comment]
@mikemcquaid Yes. I can untap some of the non-homebrew, non-caskroom taps, if those are an issue; I don't use them anymore. I've also run `brew update && brew upgrade` recently, since my `brew` was quite old (6-9 months, maybe?) for a while, so there might be an issue related to that.

```
bash-3.2$ brew tap
caskroom/cask
caskroom/fonts
caskroom/versions
goxberry/itaps
homebrew/dupes
homebrew/python
homebrew/science
homebrew/versions
homebrew/x11
optimizers/fenics
```

**Comment by goxberry** [comment]
Reproduced on another machine:

```
Geoffs-MacBook-Pro:local goxberry$ brew uses open-mpi
boost                              homebrew/science/garli             homebrew/science/mumps             homebrew/science/slepc
bsponmpi                           homebrew/science/getdp             homebrew/science/neuron            homebrew/science/sundials
fftw                               homebrew/science/gmsh              homebrew/science/nonpareil         homebrew/science/superlu_dist
konoha                             homebrew/science/gromacs           homebrew/science/p4est             homebrew/science/trilinos
homebrew/science/abinit            homebrew/science/hdf5              homebrew/science/paraview          homebrew/science/zoltan
homebrew/science/abyss             homebrew/science/hyphy             homebrew/science/parmetis          homebrew/versions/boost149
homebrew/science/alpscore          homebrew/science/hypre             homebrew/science/pastix            homebrew/versions/boost150
homebrew/science/arpack            homebrew/science/lammps            homebrew/science/petsc             optimizers/fenics/dolfin
homebrew/science/bless             homebrew/science/lis               homebrew/science/ray               optimizers/fenics/hypre-old
homebrew/science/dealii            homebrew/science/maker             homebrew/science/scalapack         optimizers/fenics/petsc-fenics
homebrew/science/elemental         homebrew/science/meme              homebrew/science/scotch
homebrew/science/exabayes          homebrew/science/mrbayes           homebrew/science/scotch5
Geoffs-MacBook-Pro:local goxberry$ brew uses mpich2
Geoffs-MacBook-Pro:local goxberry$ brew uses --installed open-mpi
homebrew/science/arpack     homebrew/science/gmsh       homebrew/science/scalapack
fftw                        homebrew/science/hdf5       homebrew/science/scotch5
```

With a recent `brew update`, currently at commit 4056cf4caf85d71968e7a87259bc5585d620432b on this machine.

**Comment by mikemcquaid** [comment]
> tl;dr: It seems like brew uses mpich2 should return something, because it's possible to use mpich2 to build other Homebrew packages that depend on MPI.

Ah, ok, I misunderstood the first bit. Unless those packages have a `depends_on "mpich2"` they won't show up here. If you want to add it to them you'll need to submit a PR to change the relevant formulae. Probably the right solution is have a `:mpi` requirement with `open-mpi` as the `default_formula` but which can find `mpich2` instead.

**Comment by goxberry** [comment]
> Unless those packages have a depends_on "mpich2" they won't show up here. If you want to add it to them you'll need to submit a PR to change the relevant formulae. 

In theory, any library that depends on MPI and uses `depends_on :mpi` should be buildable with either `open-mpi` or `mpich2`. In practice, because MPICH and Open MPI implement the standard differently, not everything can be built with either library. I agree that those formulas requiring a specific MPI implementation should use `depends_on 'open-mpi'` or `depends_on 'mpich2'`.

For those formulas that could be built with either MPICH or Open MPI, I think you're right when you say

> Probably the right solution is have a :mpi requirement with open-mpi as the default_formula but which can find mpich2 instead.

Will take a look and see if I can take a crack at a PR. My first thought is kind of hacky, but maybe it'll start a good discussion.

**Comment by tdsmith** [comment]
I believe this is already implemented.

**Comment by mikemcquaid** [comment]
My apologies; as @tdsmith pointed out this is already implemented. I can see an argument for `brew uses mpi` to work as you describe but I don't think `brew uses mpich2` can.

**Comment by goxberry** [comment]
@tdsmith Sort of. The MPI requirement is implemented. The `brew uses` command only uses `open-mpi` to list dependencies because that is the `default_formula`. 

@mikemcquaid I think that's a good possibility. I see two potential issues with your suggestion:

1) Even though both Open MPI and MPICH implement the MPI standard, sometimes, software only works with one of these implementations. While this situation should not happen, until issues are resolved with the incompatible implementation of MPI, it makes sense to express a dependency on the compatible MPI implementation. Example: For a while, boost depended on Open MPI when compiling boost with C++11 features.

2) `brew uses mpi` seems unexpected because MPI is not a package, and could miss packages expressing an explicit dependency on Open MPI or MPICH. I could see getting around user interface issues with a proper console notification, like when a user tries to install LaTeX via brew, or the various caveat notifications included with formulas.
